README.md
# TinyServerClient
Small (tiny) tcp server and client for testing.

This repository contains only two files:

    * tinyserver.py - Small multi-threaded server, with configurable port, number of clients, 
                      buffer size, and message to receive that kills the process.
                      
    * tinyclient.py - Client to use with tinyserver.py and to be able to do communication tests. 
                      It also allows you to configure the host, port and message to send.
      

These modules have been used for the execution of tests of another project that is currently underway (August / 2019), and is not finished.

Please feel free to contact me if you wish to make any suggestions or send me any comments.

If you use this code or part of it for any project of yours, I would like to hear from it.

I will be glad to know that it has been useful to you.


# Usage
There is no gui, all by command line. Different parameters can be passed, but without parameters it also works because it has set default values.

    usage: tinyserver.py [-h] [-i INTERFACEBIND] [-c MAXCLIENTS] [-p SERVERPORT]
                         [-b BUFFERSIZE] [-k KILLMESSAGE] [-v {0,1,2}]
       
    optional arguments:
      -h, --help            show this help message and exit
      -i INTERFACEBIND, --interfacebind INTERFACEBIND
                            What interface is it linked to? Default (empty) to any
                            of this host.
      -c MAXCLIENTS, --maxclients MAXCLIENTS
                            Indicates the maximum number of clients/threads.
                            Default value: 10
      -p SERVERPORT, --serverport SERVERPORT
                            Port to listen. Default value: 6666
      -b BUFFERSIZE, --buffersize BUFFERSIZE
                            Buffer size. Default value: 4096
      -k KILLMESSAGE, --killmessage KILLMESSAGE
                            Message to kill server. Default value: 'kill'
      -v {0,1,2}, --verbose {0,1,2}
                            Increase output verbosity. Default value: 1
    
    
    usage: tinyclient.py [-h] [-s HOSTSERVER] [-p PORT] [-b BUFFERSIZE]
                         [-m MESSAGE] [-v {0,1,2}]
       
    optional arguments:
      -h, --help            show this help message and exit
      -s HOSTSERVER, --hostserver HOSTSERVER
                            Host to connect. Default 127.0.0.1
      -p PORT, --port PORT  Port to connect. Default value: 6666
      -b BUFFERSIZE, --buffersize BUFFERSIZE
                            Buffer size. Default value: 4096
      -m MESSAGE, --message MESSAGE
                            Message to send to the server. Default value: 'kill'
      -v {0,1,2}, --verbose {0,1,2}
                            Increase output verbosity. Default value: 1



# References
Part of the creation of this code has been inspired by the publication of the following blog by Daniel Hnyk (@hnykda).

http://danielhnyk.cz/simple-server-client-aplication-python-3/

And from here I appreciate his comments and suggestions to my first version of the code.

# Contact
email: gabimarti (at) gmail (dot) com

twitter: @gmarti
                

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

This document is not intended to be perfect, so it is likely that it may contain some error. 
If so, I will thank you for your comments.


## [1.0.1] - 2019-08-07
### Added
- Start using "CHANGELOG.md"
- Added folder old_versions and moved older versions with subfolders here.


### Changed
- Start versioning based on the current version at 1.0.1 
- Small changes in the README.md
- In some mixed string formatting I substitute the use of % and concatenate variables with the use of format().
- I replace the double quotes with single quotes in some strings. Simply by homogeneity throughout the code.


### Removed
- Elimination of unnecessary texts in the module header of tinyserver.py and tinyclient.py


### Fixed
- Fix typos in recent README changes.





tinyclient.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# -----------------------------------------------------------------------------------------------------------
# Name:         tinyclient.py
# Purpose:      Small client program to use with tiny server.
#               Send a standard or predefined message and receive the respect from the server.
#
# Author:       Gabriel Marti Fuentes
# email:        gabimarti at gmail dot com
# GitHub:       https://github.com/gabimarti
# Created:      04/08/2019
# Version:      1.0
# License:      MIT License
# Notes:        Inspired by Daniel Hnyk's tutorial
#               http://danielhnyk.cz/simple-server-client-aplication-python-3/
#
#               Please feel free to contact me if you wish to make any suggestions or send me any comments.
#               If you use this code or part of it for any project of yours, I would like to hear from it.
#               I will be glad to know that it has been useful to you.
# -----------------------------------------------------------------------------------------------------------
#


import argparse
import socket
import sys
import threading
import time
import traceback

########################################################
# CONSTANTS
########################################################
DESCRIPTION = 'Tiny Client'
EPILOG = 'Connect, send and receive. Simply.'
DEFAULT_VERBOSE_LEVEL = 1                                   # Default verbose level.
VERBOSE_LEVEL_DESCRIPTION = ['basic',                       # Arbitrary values to adjust
                             'a few',
                             'insane info']                 # Verbose levels description
DEFAULT_HOST_CONNECT = '127.0.0.1'                          # Host to connect
DEFAULT_PORT = 6666                                         # Default port
DEFAULT_KILL_MESSAGE = 'kill'                               # Message to kill server, if no message is indicate
DEFAULT_MESSAGE_IF_EMPTY = DESCRIPTION                      # If empty message is established, replace it with this one.
DEFAULT_MAX_BUFFER_SIZE = 4096                              # Default max buffer size
ENCODING = 'utf-8'                                          # Encoding for message communication
SOCKET_TIMEOUT = 3                                          # 3 seconds are more than enough


########################################################
# VARIABLES
########################################################
host_connect = DEFAULT_HOST_CONNECT
server_port = DEFAULT_PORT
buffer_size = DEFAULT_MAX_BUFFER_SIZE
message = DEFAULT_KILL_MESSAGE
verbose = DEFAULT_VERBOSE_LEVEL                             # Verbose level
time_start = 0                                              # Simply to measure times


########################################################
# FUNCTIONS
########################################################

# Print the message if the verbose level allows
def print_verbose(msg, verbose_level, established_verbose):
    if verbose_level <= established_verbose:
        print(msg)


# Parse command line parameters
def parse_params():
    parser = argparse.ArgumentParser(description=DESCRIPTION, epilog=EPILOG)
    parser.add_argument('-s', '--hostserver', type=str, default=DEFAULT_HOST_CONNECT,
                        help='Host to connect. Default %s ' % DEFAULT_HOST_CONNECT)
    parser.add_argument('-p', '--port', type=int, default=DEFAULT_PORT,
                        help='Port to connect. Default value: %s ' % str(DEFAULT_PORT))
    parser.add_argument('-b', '--buffersize', type=int, default=DEFAULT_MAX_BUFFER_SIZE,
                        help='Buffer size. Default value: %s ' % str(DEFAULT_MAX_BUFFER_SIZE))
    parser.add_argument('-m', '--message', type=str, default=DEFAULT_KILL_MESSAGE,
                        help='Message to send to the server. Default value: \'%s\'' % str(DEFAULT_KILL_MESSAGE))
    parser.add_argument('-v', '--verbose', type=int, choices=[0, 1, 2], default=DEFAULT_VERBOSE_LEVEL,
                        help='Increase output verbosity. Default value: %s' % DEFAULT_VERBOSE_LEVEL)
    args = parser.parse_args()
    return args


# Main - Start Client
def start_client():
    global host_connect, server_port, buffer_size, message, verbose, time_start

    # Check and parse parameters
    args = parse_params()
    host_connect = args.hostserver
    server_port = args.port
    buffer_size = args.buffersize
    message = args.message
    verbose = args.verbose

    if buffer_size < 64:                                    # avoid too small buffer size
        buffer_size = DEFAULT_MAX_BUFFER_SIZE
        print_verbose('Buffer size too small. Setting to {:d} bytes'.format(buffer_size) , 1, verbose)

    if message == '':                                       # avoid empty message
        message = DEFAULT_MESSAGE_IF_EMPTY
        print_verbose('Message is empty. Setting to \'{}\''.format(message), 1, verbose)

    print('Verbose level %s ' % str(VERBOSE_LEVEL_DESCRIPTION[verbose]))
    print('Connect to Host %s' % host_connect)
    print('Server Port %d' % server_port)
    print('Buffer size %d bytes' % buffer_size)
    print('Message to send \'%s\'' % message)
    print('Starting Client ...')

    time_start = time.perf_counter()

    try:
        print_verbose('Connecting ...', 1, verbose)
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.settimeout(SOCKET_TIMEOUT)
        s.connect((host_connect, server_port))
        # Send message
        print_verbose('Sending \'{}\' ...'.format(message), 1, verbose)
        s.send(message.encode(ENCODING))                    # Encoding string to bytes
        # Wait for server response
        print_verbose('Waiting response ...', 2, verbose)
        response_bytes = s.recv(buffer_size)
        response_string = response_bytes.decode(ENCODING)   # Decode response in bytes
        print("Response from server is \'{}\'".format(response_string))
    except Exception as e:
        print('An error has occurred.')
        print('Exception : ' + str(e))
    finally:
        s.close()

    # Time measurement
    total_time = time.perf_counter() - time_start
    print('The process has happened in {:3.2f} seconds'.format(total_time))
    sys.exit()


if __name__ == '__main__':
    start_client()



tinyserver.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# -----------------------------------------------------------------------------------------------------------
# Name:         tinyserver.py
# Purpose:      Small server using threads that listens on a default port and returns a response
#               with echo of the same message received and additional information.
#
# Author:       Gabriel Marti Fuentes
# email:        gabimarti at gmail dot com
# GitHub:       https://github.com/gabimarti
# Created:      03/08/2019
# Version:      1.0
# License:      MIT License
# Notes:        Inspired by Daniel Hnyk's tutorial
#               http://danielhnyk.cz/simple-server-client-aplication-python-3/
#
#               Please feel free to contact me if you wish to make any suggestions or send me any comments.
#               If you use this code or part of it for any project of yours, I would like to hear from it.
#               I will be glad to know that it has been useful to you.
# -----------------------------------------------------------------------------------------------------------
#


import argparse
import atexit
import socket
import sys
import threading
import time
import traceback

########################################################
# CONSTANTS
########################################################
DESCRIPTION = 'Tiny Server'
EPILOG = 'Listen and respond. Simply.'
DEFAULT_MAXCLIENTS = 10                                     # Aka MAXTHREADS, every Client is a thread
DEFAULT_VERBOSE_LEVEL = 1                                   # Default verbose level
VERBOSE_LEVEL_DESCRIPTION = ['basic',
                             'a few',
                             'insane info']                 # Verbose levels description
DEFAULT_HOST_BIND = ''                                      # Bind to all ports
DEFAULT_PORT = 6666                                         # Default port
DEFAULT_KILL_MESSAGE = 'kill'                               # If this message is received, the server will end.
DEFAULT_MAX_BUFFER_SIZE = 4096                              # Default max buffer size
ENCODING = 'utf-8'                                          # Encoding for message communication
SHUTDOWN_RESPONSE = 'Shutting down the server'              # Shutdown response to client


########################################################
# VARIABLES
########################################################
host_bind = DEFAULT_HOST_BIND
max_clients = DEFAULT_MAXCLIENTS
server_port = DEFAULT_PORT
max_buffer_size = DEFAULT_MAX_BUFFER_SIZE
kill_message = DEFAULT_KILL_MESSAGE
verbose = DEFAULT_VERBOSE_LEVEL                             # Verbose level
thread_counter = 0                                          # Total executed threads / clients
thread_active_counter = 0                                   # active threads / clients
thread_client_list = []                                     # List of active threads
time_start = 0                                              # For Timer counter
shutdown = False                                            # It will be True when the server has to be closed.
server = None                                               # Global server socket


########################################################
# FUNCTIONS
########################################################

# Print the message if the verbose level allows
def print_verbose(msg, verbose_level, established_verbose):
    if verbose_level <= established_verbose:
        print(msg)


# Kill all client active sockets
def kill_client_sockets():
    global thread_client_list
    print_verbose('Killing clients ...', 1, verbose)
    for client in thread_client_list:
        try:
            client.shutdown(socket.SHUT_RDWR)
            client.close()
        except socket.error:
            pass                    # socket already closed. don't worry


# Client input string processing
def do_client_input_string_processing(input_string, verbose):
    print_verbose('Processing '+str(input_string), 1, verbose)
    # Reverse message
    input_processed = input_string[::-1]
    return input_processed


# Client thread for each new connection
def client_thread(conn, ip, port, buffer_size, verbose, kill_message):
    global shutdown, thread_active_counter, thread_client_list, server

    # Lock for shared counters operations
    lock = threading.Lock()
    lock.acquire()
    thread_active_counter += 1
    thread_client_list.append(conn)
    lock.release()

    print_verbose('Total clients {} | Current active clients {}'.format(thread_counter, thread_active_counter),
                  1, verbose)

    # the input is in bytes, so decode it
    input_from_client_bytes = conn.recv(max_buffer_size)
    if not input_from_client_bytes:
        print_verbose('No data. from {}:{}'.format(ip, port), 1, verbose)

    # max_buffer_size indicates how big the message can be
    # this is test if it's sufficiently big
    siz = sys.getsizeof(input_from_client_bytes)
    if siz >= buffer_size:
        print_verbose('The length of input is probably too long: {}'.format(siz), 1, verbose)

    # decode input and strip the end of line
    input_from_client = input_from_client_bytes.decode(ENCODING).rstrip()

    # check if message is received to kill server
    if input_from_client == kill_message:
        print_verbose('Killing message received \'{}\''.format(input_from_client), 1, verbose)
        # response shutdown to client
        res = SHUTDOWN_RESPONSE
        res_encoded = res.encode(ENCODING)
        conn.sendall(res_encoded)
        shutdown = True
        try:
            print_verbose('Shutdown server from Client thread', 1, verbose)
            server.close()
        except Exception as e:
            print_verbose('Can\'t close Server from client thread', 1, verbose)
            print_verbose('Exception {}'.format(e), 1, verbose)

        kill_client_sockets()
    else:
        res = do_client_input_string_processing(input_from_client, verbose)
        print_verbose('Result of processing {} is: {}'.format(input_from_client, res), 1, verbose)

        res_encoded = res.encode(ENCODING)              # encode the result string
        try:
            conn.sendall(res_encoded)                   # send it to client
            conn.close()                                # close connection
        except Exception as e:
            print_verbose('Socket Error: {:s} '.format(e), 1, verbose)

    # Lock for shared counters operations
    lock.acquire()
    thread_active_counter -= 1
    thread_client_list.remove(conn)
    lock.release()

    print_verbose('Connection ' + ip + ':' + port + " ended", 1, verbose)


# Parse command line parameters
def parse_params():
    parser = argparse.ArgumentParser(description=DESCRIPTION, epilog=EPILOG)
    parser.add_argument('-i', '--interfacebind', type=str, default=DEFAULT_HOST_BIND,
                        help='What interface is it linked to? Default (empty) to any of this host.')
    parser.add_argument('-c', '--maxclients', type=int, default=DEFAULT_MAXCLIENTS,
                        help='Indicates the maximum number of clients/threads. Default value: %s '
                             % str(DEFAULT_MAXCLIENTS))
    parser.add_argument('-p', '--serverport', type=int, default=DEFAULT_PORT,
                        help='Port to listen. Default value: %s ' % str(DEFAULT_PORT))
    parser.add_argument('-b', '--buffersize', type=int, default=DEFAULT_MAX_BUFFER_SIZE,
                        help='Buffer size. Default value: %s ' % str(DEFAULT_MAX_BUFFER_SIZE))
    parser.add_argument('-k', '--killmessage', type=str, default=DEFAULT_KILL_MESSAGE,
                        help='Message to kill server. Default value: \'%s\'' % str(DEFAULT_KILL_MESSAGE))
    parser.add_argument('-v', '--verbose', type=int, choices=[0, 1, 2], default=DEFAULT_VERBOSE_LEVEL,
                        help='Increase output verbosity. Default value: %s' % DEFAULT_VERBOSE_LEVEL)
    args = parser.parse_args()
    return args


# Executes on program close
def on_close_program():
    global time_start

    totaltime = time.perf_counter() - time_start
    print('The server has been running for {} seconds.'.format(totaltime))


# Main - Start Server
def start_server():
    global host_bind, max_clients, server_port, max_buffer_size, kill_message, verbose, time_start
    global shutdown, thread_counter, server

    # Handler to do actions when application is closed
    atexit.register(on_close_program)

    # Check and parse parameters
    args = parse_params()
    host_bind = args.interfacebind
    max_clients = args.maxclients
    server_port = args.serverport
    max_buffer_size = args.buffersize
    kill_message = args.killmessage
    verbose = args.verbose

    if max_clients < 1:                     # avoid zero division
        max_clients = DEFAULT_MAXCLIENTS

    print('Verbose level %s ' % str(VERBOSE_LEVEL_DESCRIPTION[verbose]))
    print('Host bind %s' % host_bind)
    print('Max %d Clients / Threads ' % max_clients)
    print('Server Port %d' % server_port)
    print('Buffer size %d bytes' % max_buffer_size)
    print('Kill message \'%s\'' % kill_message)
    print('Starting Server ...')
    time_start = time.perf_counter()

    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    print_verbose('Socket created', 1, verbose)

    # Bind socket to local host and port
    try:
        server.bind((host_bind, server_port))
        print_verbose('Socket bind completed', 2, verbose)
    except socket.error as msg:
        print('Bind failed. Error Code : ' + str(msg[0]) + ' Message ' + msg[1])
        print('Sys Error Info : ' + str(sys.exc_info()))
        sys.exit()

    # Start listening on socket
    server.listen(max_clients)
    print_verbose('Socket now listening', 2, verbose)

    # Now keep talking with the client while no shutdown message is received
    while not shutdown:
        # Wait to accept a connection - blocking call
        try:
            conn, addr = server.accept()
        except:
            break                                   # Possibly closed server

        ip, port = str(addr[0]), str(addr[1])
        print('Accepting connection from ' + ip + ':' + port)
        thread_counter += 1
        try:
            client = threading.Thread(target=client_thread,
                                      args=(conn, ip, port, max_buffer_size, verbose, kill_message))
            client.start()
        except Exception as e:
            print('WTF Error! '+str(e))
            traceback.print_exc()

    # Shutting down server
    try:
        print("Shutdown server")
        server.shutdown(socket.SHUT_RDWR)
        server.close()
    except Exception as e:
        print_verbose("Server already closed from client thread", 1, verbose)
        print_verbose("Exception {}".format(e), 2, verbose)

    print_verbose('At shutdown, total clients {} | Current active {}'.format(thread_counter,
                                                                             thread_active_counter),
                  1, verbose)
    sys.exit(1)


if __name__ == '__main__':
    start_server()

